### ISSUE TITLE
conflicting SHA1 symbol

### ISSUE THREAD

**Issue by diehard2** [launchdarkly/c-server-sdk#77]
There's a conflicting symbol name with OpenSSL's libcrypto if the sdk is built as a static library

/opt/rh/devtoolset-11/root/usr/libexec/gcc/x86_64-redhat-linux/11/ld.gold: error: vcpkg_installed/x64-linux/lib/libcrypto.a(libcrypto-lib-sha1_one.o): multiple definition of 'SHA1'
/opt/rh/devtoolset-11/root/usr/libexec/gcc/x86_64-redhat-linux/11/ld.gold: vcpkg_installed/x64-linux/lib/libldserverapi.a(sha1.c.o): previous definition here

I figured I would probably have better luck asking for a solution here than in openssl. redefining the symbol  looks (at best) to be a dicey approach. Is there any sort of workaround, or could that maybe get renamed?

**Comment by cwaldren-ld** [comment]
Hi @diehard2 thanks for the report, can you please post the compiler arguments/invocation you are using when you encounter this issue?

**Comment by diehard2** [comment]
Its in a huge project right now. I can try to create a smaller reproducer.

**Comment by diehard2** [comment]
@cwaldren-ld sorry for not getting back to this, I switched to dynamic, but I finally have a reproducer since I would really prefer static linkage. I created my own version of 2.9.2 and I see the same issue.

vcpkg.json
```
{
    "name": "third-party",
    "version": "1.0.0",
    "dependencies": [
	"launch-darkly-server",
	"cryptopp"
    ],
    "builtin-baseline": "e85cafa30def190eca0dc1413bf8c04e434af567"
}
```

CMakeFiles.txt
```
cmake_minimum_required(VERSION 3.16)

set(CMAKE_TOOLCHAIN_FILE ${CMAKE_CURRENT_SOURCE_DIR}/vcpkg/scripts/buildsystems/vcpkg.cmake)

project(test)
find_package(cryptopp CONFIG REQUIRED)
find_package(launch-darkly-server CONFIG REQUIRED)
find_package(Threads)
add_executable(test main.cxx)
target_link_libraries(test ldserverapi::ldserverapi cryptopp::cryptopp)

```
main.cxx
``` 
#include <iostream>
#include <launchdarkly/api.h>
#include <launchdarkly/client.h> // Add this line
#include <cryptopp/sha.h>
#include <cryptopp/hex.h>

int main() {
    // Instantiate LaunchDarkly server client
    LDConfig* config = LDConfigNew("YOUR_SDK_KEY");
    LDClient* client = LDClientInit(config, 0);
    CryptoPP::SHA1 sha1;
    std::string message = "Hello, World!";
    CryptoPP::byte digest[CryptoPP::SHA1::DIGESTSIZE];
    sha1.CalculateDigest(digest, (const CryptoPP::byte*)message.c_str(), message.length());

    // Convert digest to hexadecimal string
    std::string encodedDigest;
    CryptoPP::StringSource(digest, sizeof(digest), true,
        new CryptoPP::HexEncoder(
            new CryptoPP::StringSink(encodedDigest)
        )
    );
    std::cout << "SHA1 Digest: " << encodedDigest << std::endl;
    return 0;
}

```
the error - 
```
[2/2] Linking CXX executable test
FAILED: test 
: && /usr/bin/c++     CMakeFiles/test.dir/main.cxx.o  -o test  vcpkg_installed/x64-linux/debug/lib/libldserverapi.a  vcpkg_installed/x64-linux/debug/lib/libcryptopp.a  vcpkg_installed/x64-linux/lib/libcurl.a  vcpkg_installed/x64-linux/lib/libssl.a  vcpkg_installed/x64-linux/lib/libcrypto.a  -ldl  vcpkg_installed/x64-linux/lib/libssl.a  vcpkg_installed/x64-linux/lib/libcrypto.a  -ldl  vcpkg_installed/x64-linux/lib/libz.a  vcpkg_installed/x64-linux/lib/libpcre.a  -lm  -lpthread && :
/usr/bin/ld: vcpkg_installed/x64-linux/lib/libcrypto.a(libcrypto-lib-sha1_one.o): in function `SHA1':
sha1_one.c:(.text+0xa0): multiple definition of `SHA1'; vcpkg_installed/x64-linux/debug/lib/libldserverapi.a(sha1.c.o):/home/myname/deleteme/ld_error/vcpkg/buildtrees/launch-darkly-server/src/274bd6b472-130d2b5c68.clean/sha1.c:286: first defined here
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.

```

**Comment by diehard2** [comment]
I fixed it in my build with this patch for CLIBS and then making the corresponding change in the sdk

clibs_patch.diff


```
diff --git a/sha1.c b/sha1.c
index 76cd6ca..b30617c 100644
--- a/sha1.c
+++ b/sha1.c
@@ -279,7 +279,7 @@ void SHA1Final(
     memset(&finalcount, '\0', sizeof(finalcount));
 }
 
-void SHA1(
+void SHA1_CLIBS(
     char *hash_out,
     const char *str,
     uint32_t len)
diff --git a/sha1.h b/sha1.h
index f492009..b7bf5a0 100644
--- a/sha1.h
+++ b/sha1.h
@@ -40,7 +40,7 @@ void SHA1Final(
     SHA1_CTX * context
     );
 
-void SHA1(
+void SHA1_CLIBS(
     char *hash_out,
     const char *str,
     uint32_t len);
```

**Comment by cwaldren-ld** [comment]
I think it would be reasonable for us to include this patch in the SDK's source and then apply it after the `FetchContent` step. 

As a side note - our development effort in the C++ and C world is now focused on the [C++ Server-side SDK with C bindings](https://github.com/launchdarkly/cpp-sdks). This SDK is deprecated and will go EOL on 2024-12-04.

**Comment by diehard2** [comment]
Thanks Casey, much appreciated. The mono-repo currently doesn't have a package in vcpkg so I can't give it a try at the moment. But I'll switch over whenever they get around to it (I'm not sure who maintains all of those)

**Comment by cwaldren-ld** [comment]
Hi @diehard2, [2.9.3](https://github.com/launchdarkly/c-server-sdk/releases/tag/2.9.3) patches the `clibs/sha1` library to rename the symbol. 

Thanks again for the report and sorry for the long wait.

**Comment by cwaldren-ld** [comment]
Hi @diehard2, I'm going to close this now as it has been patched in 2.9.3.

I see vcpkg ended up [merging a change to patch it directly](https://github.com/microsoft/vcpkg/pull/35901), which could probably be reverted.
